### PR TITLE
EES-2031 Add quantile options for map charts

### PR DESCRIPTION
**Pull request by ntsim**
This PR adds options for users to build map charts using either equal intervals (groups of equally sized ranges) or quantiles (groups with an equal number of features).

With equal intervals:

image

With quantiles:

image

## New 'Map configuration' tab in chart builder

A new 'Map configuration' tab has been introduced as part of the chart builder UI:

image

We have importantly moved the boundary levels selector into this tab (from previously being in 'Chart configuration') so that all map-specific configuration is grouped together.

As part of this new tab, the user can specify the number of data classes (a.k.a. features, groups, etc), giving them additional control over how the data will be classified.

## Related changes

- Added `color2k` as a dependency. This provides a bunch of useful colour related util functions that replace homebrew functions such as `lighten`.
- Map colours have been slightly tweaked to be lighter. We now start from a max of `#fff` (white) and use the desired colour as the last group's colour. Previously, most group colours would quickly move towards really dark shades making the overall palette seem a little dark.

## Other changes

- Added visually hidden 'Error: ' prefixes to `ErrorMessage` contents. This is for parity with the GDS pattern for validation error messages.
- Various fixing of math/number functions to handle potential edge cases around NaN, Infinity and empty strings.
- Removed redundant `numberWithCommas` function in favour of `formatPretty`.

### CODE REVIEW

File: src/explore-education-statistics-common/src/modules/charts/components/MapBlockInternal.tsx [hunk 1]
```diff
@@ -726,6 +504,97 @@ export const MapBlockInternal = ({
...
+  groups: number;
+  selectedDataSetConfig: DataSetCategoryConfig;
+}): {
+  features: MapFeatureCollection;
+  dataClasses: LegendDataClass[];
+} {
+  const selectedDataSetKey = selectedDataSetConfig.dataKey;
+  const {
+    unit,
+    decimalPlaces: dataSetDecimalPlaces,
```
Comment (amyb-hiveit) [inline src/explore-education-statistics-common/src/modules/charts/components/MapBlockInternal.tsx:526]: I don't think dataSetDecimalPlaces is actually being used if it's set, should it be passed to generateLegendDataClasses?

File: src/explore-education-statistics-common/src/modules/charts/components/utils/generateLegendDataClasses.ts [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+      if (dp > implicitDecimalPlaces) {
+        implicitDecimalPlaces = dp;
+      }
+    }
+
+    return value;
+  });
+
+  let decimalPlaces =
```
Comment (amyb-hiveit) [inline src/explore-education-statistics-common/src/modules/charts/components/utils/generateLegendDataClasses.ts:64]: defaultMaxDecimalPlaces isn't being respected here, I'm getting 5 decimal places on this branch where on dev it's limited to 2. Also, the colours are a lot duller, is that intended?

map-pr
map-dev

File: src/explore-education-statistics-common/src/modules/charts/components/__tests__/MapBlockInternal.test.tsx [hunk 3]
```diff
@@ -59,28 +59,28 @@ describe('MapBlockInternal', () => {
...
-    expect(legendItems[0]).toHaveTextContent('3.0% to 3.2%');
-    expect(legendItems[1]).toHaveTextContent('3.3% to 3.4%');
-    expect(legendItems[2]).toHaveTextContent('3.5% to 3.6%');
-    expect(legendItems[3]).toHaveTextContent('3.7% to 3.8%');
-    expect(legendItems[4]).toHaveTextContent('3.9% to 4.0%');
+    expect(legendItems[0]).toHaveTextContent('3.0% to 3.1%');
+    expect(legendItems[1]).toHaveTextContent('3.2% to 3.3%');
+    expect(legendItems[2]).toHaveTextContent('3.4% to 3.5%');
+    expect(legendItems[3]).toHaveTextContent('3.6% to 3.7%');
+    expect(legendItems[4]).toHaveTextContent('3.8% to 4.0%');
```
Comment (amyb-hiveit) [inline src/explore-education-statistics-common/src/modules/charts/components/__tests__/MapBlockInternal.test.tsx:74]: Why have these changed, it doesn't look like the test data has?

File: src/explore-education-statistics-common/src/modules/charts/components/__tests__/MapBlockInternal.test.tsx [hunk 4]
```diff
@@ -59,28 +59,28 @@ describe('MapBlockInternal', () => {
...
-    expect(legendColours[0].style.backgroundColor).toBe('rgb(134, 188, 255)');
-    expect(legendColours[1].style.backgroundColor).toBe('rgb(113, 158, 255)');
-    expect(legendColours[2].style.backgroundColor).toBe('rgb(92, 128, 214)');
-    expect(legendColours[3].style.backgroundColor).toBe('rgb(71, 99, 165)');
-    expect(legendColours[4].style.backgroundColor).toBe('rgb(49, 69, 115)');
+    expect(legendColours[0].style.backgroundColor).toBe('rgb(218, 224, 237)');
+    expect(legendColours[1].style.backgroundColor).toBe('rgb(181, 193, 219)');
+    expect(legendColours[2].style.backgroundColor).toBe('rgb(145, 161, 201)');
+    expect(legendColours[3].style.backgroundColor).toBe('rgb(108, 130, 183)');
+    expect(legendColours[4].style.backgroundColor).toBe('rgb(71, 99, 165)');
```
Comment (amyb-hiveit) [inline src/explore-education-statistics-common/src/modules/charts/components/__tests__/MapBlockInternal.test.tsx:83]: Not related to this test specifically, but will these colour changes affect existing maps? If so, is that ok?

File: src/explore-education-statistics-common/src/modules/charts/components/utils/__tests__/generateLegendDataClasses.test.ts [hunk 5]
```diff
@@ -0,0 +1,1145 @@
+import generateLegendDataClasses, {
+  LegendDataClass,
+} from '@common/modules/charts/components/utils/generateLegendDataClasses';
+
+describe('generateLegendDataClasses', () => {
```
Comment (amyb-hiveit) [inline src/explore-education-statistics-common/src/modules/charts/components/utils/__tests__/generateLegendDataClasses.test.ts:5]: This needs tests for when the values have more than 2 decimal places and if decimalPlaces is passed in.

**Review by amyb-hiveit — CHANGES_REQUESTED** [review]

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/ChartMapConfiguration.tsx [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+  const validationSchema = useMemo(() => {
+    let schema = Yup.object<FormValues>({
+      dataClassification: Yup.string()
+        .required('Choose a data classification')
+        .oneOf(['EqualIntervals', 'Quantiles']) as StringSchema<
+        DataClassification
+      >,
+      dataClasses: Yup.number()
+        .required('Enter the number of data classes')
+        .min(1, 'The number of data classes must be at least 1'),
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/ChartMapConfiguration.tsx:62]: I've set a max of 100. Tbh, this is a bit ridiculous anyway, but it doesn't crash at least!

Btw, I've also changed the terminology from 'data classes' to 'data groups' on the suggestion of Cam.

File: src/explore-education-statistics-common/src/modules/charts/components/MapBlockInternal.tsx [hunk 7]
```diff
@@ -726,6 +504,97 @@ export const MapBlockInternal = ({
...
+    decimalPlaces: number;
+  }>(
+    (acc, category) => {
+      const value = category.dataSets[selectedDataSetKey]?.value;
+
+      if (Number.isFinite(value)) {
+        acc.values.push(value);
+
+        if (typeof dataSetDecimalPlaces !== 'number') {
+          const dp = countDecimalPlaces(value.toString()) ?? 0;
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/explore-education-statistics-common/src/modules/charts/components/MapBlockInternal.tsx:544]: Good point, it doesn't need to :smile: There was a lot of churn with this code whilst I was re-working it and must have missed some of this stuff. Thanks!

I've done a few things:
- Renamed `generateLegendDataClasses` -> `generateLegendDataGroups` to reflect change in terminology from 'data classes' to 'data groups' (as mentioned in another comment).
- Decimal places get calculated inside of `generateLegendDataGroups`
- We don't bother converting the numbers to strings anymore

File: src/explore-education-statistics-common/src/modules/charts/components/utils/generateLegendDataClasses.ts (hunk 2 above)
Comment (ntsim) [reply src/explore-education-statistics-common/src/modules/charts/components/utils/generateLegendDataClasses.ts:64]: Oops, there was a mistake there! Should now respect 2 d.p and there's some tests for it as well.

In terms of the colours - this is mostly intentional. The previous iteration wasn't really scaled properly as it was using some hand-picked values to get a 'nice' palette based on 5 groups. This involved a bunch of manual trial and error so it really wasn't a practical way to generate colours for a user-defined number of groups.

I've currently scaled it from white (the top-end) to the target colour (bottom-end), so this tends towards lighter tones as the range is typically wider. 

Here are some variations where we scale the colour differently:

Target colour + 60% lightness at the top-end:

image

Target colour + 50% lightness at the top-end:

image

Target colour + 40% lightness at the top-end:

image

Not sure if you have any preferences?

File: src/explore-education-statistics-common/src/modules/charts/components/__tests__/MapBlockInternal.test.tsx (hunk 4 above)
Comment (ntsim) [reply src/explore-education-statistics-common/src/modules/charts/components/__tests__/MapBlockInternal.test.tsx:83]: Colour changes will indeed existing maps. I'd say this acceptable.

File: src/explore-education-statistics-common/src/modules/charts/components/__tests__/MapBlockInternal.test.tsx (hunk 3 above)
Comment (ntsim) [reply src/explore-education-statistics-common/src/modules/charts/components/__tests__/MapBlockInternal.test.tsx:74]: This is because the rounding algorithm changed slightly. There were some issues with the previous version as it didn't handle some edge-cases well (particularly when working with only integer data) and groups could become a little uneven in size.

We now go with an approach of trying to make all groups besides the last group the same size. Unfortunately, this last group will typically end up being smaller/larger than the other groups as a result. This can feel a little random depending on how the numbers round.

File: src/explore-education-statistics-common/src/modules/charts/components/utils/__tests__/generateLegendDataClasses.test.ts (hunk 5 above)
Comment (ntsim) [reply src/explore-education-statistics-common/src/modules/charts/components/utils/__tests__/generateLegendDataClasses.test.ts:5]: Good catch, added!

File: src/explore-education-statistics-common/src/modules/charts/components/utils/generateLegendDataClasses.ts (hunk 2 above)
Comment (amyb-hiveit) [reply src/explore-education-statistics-common/src/modules/charts/components/utils/generateLegendDataClasses.ts]: I'd make it a bit brighter as it's very grey currently. Maybe + 50% lightness?

File: src/explore-education-statistics-common/src/modules/charts/components/MapBlockInternal.tsx [hunk 8]
```diff
@@ -726,6 +503,81 @@ export const MapBlockInternal = ({
...
+  dataGroups: number;
+  selectedDataSetConfig: DataSetCategoryConfig;
+}): {
+  features: MapFeatureCollection;
+  dataGroups: LegendDataGroup[];
+} {
+  const selectedDataSetKey = selectedDataSetConfig.dataKey;
+  const {
+    unit,
+    decimalPlaces: dataSetDecimalPlaces,
```
Comment (amyb-hiveit) [inline src/explore-education-statistics-common/src/modules/charts/components/MapBlockInternal.tsx:525]: Really minor, but don't you really need to rename this to dataSetDecimalPlaces now.

**Review by amyb-hiveit — APPROVED** [review]
Looks good - just one minor comment and I'll leave it up to you to decide about the colour brightness.
